compiler/rustc_session/src/session.rs

```diff
@@ -1187,7 +1187,12 @@ fn validate_commandline_args_with_session_available(sess: &Session) {
 
     // Sanitizers can only be used on platforms that we know have working sanitizer codegen.
     let supported_sanitizers = sess.target.options.supported_sanitizers;
-    let unsupported_sanitizers = sess.opts.unstable_opts.sanitizer - supported_sanitizers;
+    let mut unsupported_sanitizers = sess.opts.unstable_opts.sanitizer - supported_sanitizers;
+    // Niche: if `fixed-x18`, or effectively switching on `reserved-x18` flag, is enabled
+    // we should allow Shadow Call Stack sanitizer.
+    if sess.opts.unstable_opts.fixed_x18 && sess.target.arch == "aarch64" {
```

Comment:
This only checks `fixed-x18` is passed as opts IIUC? What if the target specify `reserved-x18` in their `features`?

Reply:
We can actually.

Reply:
Then the sanitizer is listed as supported in the target config, so it will already be in `supported_sanitizers`.

Reply:
A search showed that compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_ohos.rs has reserve-x18 set, where it doesn't set `SanitizerSet::SHADOWCALLSTACK` as supported sanitizers.

Reply:
A quick search in `ohos` revealed no support of SCS. At least not with their musl library.

Reply:
One question. How shall we handle with `aarch64-unknown-none`? The fact that whether the runtime on which the code would be executed on has SCS support is unknown to us.

The single most important use case that this patch enables is to allow code targeting `aarch64-unknown-none` to be instrumented with SCS sanitizer.

Reply:
I mean, some targets are going to support SCS only on a bring-your-own-runtime basis. We can't know up front if that's the case or not.

Reply:
Okay, I would propose to stick with the wording that user discretion is required.

In fact, checking on LLVM SCS instrumentation, it looks like it demands very little from the runtime. If necessary, we can put a link to the bionic code, so that the user will have an idea what the runtime should subscribe to.